bionic/datatypes.py

```diff
@@ -49,48 +51,54 @@ def __repr__(self):
         return f"Task({self.keys!r}, {self.dep_keys!r})"
 
 
-class Query(object):
+@attr.s(frozen=True)
+class Query:
     """
     Represents a request for a specific entity value.
     """
 
-    def __init__(self, task_key, protocol, provenance):
-        self.task_key = task_key
-        self.dnode = task_key.dnode
-        self.case_key = task_key.case_key
-        self.protocol = protocol
-        self.provenance = provenance
+    task_key = attr.ib()
+    case_key = attr.ib()
```

Comment:
I don't think you need this case_key since you have it as a property

Reply:
You're right -- thanks!